homeassistant/components/unifiprotect/media_source.py

```diff
@@ -0,0 +1,597 @@
+
+            source.children = await self._build_events(**args)  # type: ignore[arg-type,assignment]
+
+        return source
+
+    async def _build_month(
+        self,
+        data: ProtectData,
+        camera_id: str,
+        event_type: SimpleEventType,
+        start: date,
+        build_children: bool = False,
+    ) -> BrowseMediaSource:
+        """Build media source for events for a given month."""
+
+        base_id = f"{data.api.bootstrap.nvr.id}:{camera_id}:{event_type.name.lower()}"
+
+        source = BrowseMediaSource(
+            domain=DOMAIN,
+            identifier=f"{base_id}:month:{start.year}:{start.month}",
+            media_class=MEDIA_CLASS_DIRECTORY,
+            media_content_type=VIDEO_FORMAT,
+            title=f"{start.strftime('%B %Y')}",
+            can_play=False,
+            can_expand=True,
+            children_media_class=MEDIA_CLASS_VIDEO,
+        )
+
+        if build_children:
+            start_dt = datetime(
+                year=start.year,
+                month=start.month,
+                day=1,
+                hour=0,
+                minute=0,
+                second=0,
+                tzinfo=_timezone(self.hass),
+            )
+            if start_dt.month < 12:
+                end_dt = start_dt.replace(month=start_dt.month + 1)
+            else:
+                end_dt = start_dt.replace(year=start_dt.year + 1, month=1)
+
+            args = {
+                "data": data,
+                "start": start_dt,
+                "end": end_dt,
+                "reserve": False,
+            }
+            if camera_id != "all":
+                args["camera_id"] = camera_id
+            if event_type != SimpleEventType.ALL:
+                args["event_type"] = SimpleEventType.get_event_type(event_type)
+
+            source.children = await self._build_events(**args)  # type: ignore[arg-type,assignment]
+
+        return source
+
+    async def _build_events_type(
+        self,
+        data: ProtectData,
+        camera_id: str,
+        event_type: SimpleEventType,
+        build_children: bool = False,
+    ) -> BrowseMediaSource:
+        """Build media source for a specific event type."""
+
+        base_id = f"{data.api.bootstrap.nvr.id}:{camera_id}:{event_type.name.lower()}"
+
+        source = BrowseMediaSource(
+            domain=DOMAIN,
+            identifier=base_id,
+            media_class=MEDIA_CLASS_DIRECTORY,
+            media_content_type=VIDEO_FORMAT,
+            title=event_type.value.title(),
+            can_play=False,
+            can_expand=True,
+            children_media_class=MEDIA_CLASS_VIDEO,
+        )
+
+        if build_children:
+            source.children = [
+                await self._build_day(data, camera_id, event_type, 1),
+                await self._build_day(data, camera_id, event_type, 7),
+                await self._build_day(data, camera_id, event_type, 30),
+            ]
```

Comment:
Can we gather here?

Reply:
So, the `await` on most of these is misleading here. It _can_ be an async function if `build_children=True` is passed into all of the children methods and it actually fetches the events. Otherwise, it only builds one level down synchronously. So `gather` would not offer any performance benefits here.

I could still switch it to gather though for convention sake, but does gather maintain output order?

Reply:
The order of result values corresponds to the order of awaitables in aws.